implement EnqueueExtensions interface in serviceaffinity

Huang-Wei opened the pull request:
#### What type of PR is this?

/kind feature
/sig scheduling

#### What this PR does / why we need it:

Implement EnqueueExtensions interface in serviceaffinity. See the inline comments for more details.

#### Which issue(s) this PR fixes:

Part of #94009.

#### Special notes for your reviewer:

I chose to opt-out benchmark testing.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Huang-Wei committed:
implement EnqueueExtensions interface in serviceaffinity

Huang-Wei commented:
/assign @ahg-g @alculquicondor

Huang-Wei commented:
/retest

yangjunmyfm192085 commented:
/test pull-kubernetes-e2e-kind

yangjunmyfm192085 commented:
/test pull-kubernetes-e2e-kind

alculquicondor reviewed (commented):
Can you remove Service events from the "all events" list?

Huang-Wei committed:
unregister Service events if a plugin doesn't implement EventsToRegister()

Huang-Wei commented:
> Can you remove Service events from the "all events" list?

Added a commit to cover this. (also added an integration test to ensure ServiceAffinity still functions well)

alculquicondor commented on test/integration/scheduler/queue_test.go:
```diff
@@ -0,0 +1,150 @@
+	}
+
+	// Pop Pod2 out. It should be unschedulable.
+	podInfo := testCtx.Scheduler.NextPod()
```
should we expose scheduleOne instead?

Huang-Wei replied:
In very rare cases (like this one) we need that, so probably no.

alculquicondor replied:
the same could be said about `NextPod` though.

Huang-Wei replied:
🤷

Huang-Wei commented:
/retest

alculquicondor commented:
/lgtm
